### PR TITLE
[VPlan] Ignore incoming values with constant false mask.

### PR DESCRIPTION
**Pull request by fhahn**
Ignore incoming values with constant false masks when trying to simplify VPBlendRecipes.

As a follow-on optimization, we should also be able to drop all incoming values with false masks by creating a new VPBlendRecipe with those operands dropped.

### CODE REVIEW

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp [hunk 1]
```diff
@@ -884,18 +884,19 @@ void VPlanTransforms::clearReductionWrapFlags(VPlan &Plan) {
...
 /// Try to simplify recipe \p R.
 static void simplifyRecipe(VPRecipeBase &R, VPTypeAnalysis &TypeInfo) {
+  using namespace llvm::VPlanPatternMatch;
   // Try to remove redundant blend recipes.
   if (auto *Blend = dyn_cast<VPBlendRecipe>(&R)) {
     VPValue *Inc0 = Blend->getIncomingValue(0);
     for (unsigned I = 1; I != Blend->getNumIncomingValues(); ++I)
-      if (Inc0 != Blend->getIncomingValue(I))
+      if (Inc0 != Blend->getIncomingValue(I) &&
+          !match(Blend->getMask(I), m_False()))
```
Comment (ayalz) [inline llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:893]: This is fine, but could be more general: first filter all operands having false masks by replacing the original blend with a simpler blend, then apply the existing check if all operands are the same? May deserve additional tests.

**Review by ayalz — COMMENTED** [review]
Looks good to me, adding a couple of thoughts.

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 1 above)
Comment (ayalz) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:893]: (Ah, mentioned in the commit message as follow-up, very well.)

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h [hunk 2]
```diff
@@ -67,14 +67,17 @@ struct specific_intval {
...
   }
 };
 
-inline specific_intval m_SpecificInt(uint64_t V) {
-  return specific_intval(APInt(64, V));
+inline specific_intval<0> m_SpecificInt(uint64_t V) {
+  return specific_intval<0>(APInt(64, V));
 }
 
+inline specific_intval<1> m_False() { return specific_intval<1>(APInt(64, 0)); }
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h:82]: Added an assert before checking `APInt::isSameValue(CI->getValue(), Val)`, thanks!

File: llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp (hunk 1 above)
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanTransforms.cpp:893]: Will do separately, thanks!

**Review by ayalz — APPROVED** [review]
LGTM.

Long-term, optimizing masks may be addressed when generated by a VPlan2VPlan Predicator, rather than by a separate subsequent redundancy eliminator.

File: llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h [hunk 3]
```diff
@@ -52,7 +52,7 @@ template <typename Class> struct bind_ty {
 
 /// Match a specified integer value or vector of all elements of that
 /// value.
```
[the remark this answers is not in the document]
Comment (fhahn) [reply llvm/lib/Transforms/Vectorize/VPlanPatternMatch.h]: Added a comment and also checked that there are no warnings with in release builds w/o assertions (at least with clang)

**Pull request merged**